pkg/net/security/handshake/connection_handshake.go

```diff
@@ -84,65 +86,77 @@ type Act3Message struct {
 // initiatorAct1 represents the state of the initiator in the first act of the
 // handshake protocol.
 type initiatorAct1 struct {
-	nonce1 uint64
+	nonce1    uint64
+	protocol1 string
 }
 
 // InitiateHandshake function allows to initiate a handshake by creating
 // and initializing a state machine representing initiator in the first round
 // of the handshake, ready to execute the protocol.
-func InitiateHandshake() (*initiatorAct1, error) {
+func InitiateHandshake(protocol string) (*initiatorAct1, error) {
 	nonce1, err := randomNonce()
 	if err != nil {
 		return nil, fmt.Errorf("could not initiate the handshake: [%v]", err)
 	}
 
-	return &initiatorAct1{nonce1}, nil
+	return &initiatorAct1{nonce1, protocol}, nil
 }
 
 // Message returns the message sent by initiator to the responder in the first
 // act of the handshake protocol.
 func (ia1 *initiatorAct1) Message() *Act1Message {
-	return &Act1Message{nonce1: ia1.nonce1}
+	return &Act1Message{nonce1: ia1.nonce1, protocol1: ia1.protocol1}
 }
 
 // Next performs a state transition and returns initiator in a state ready to
 // execute the second act of the handshake protocol.
 func (ia1 *initiatorAct1) Next() *initiatorAct2 {
-	return &initiatorAct2{nonce1: ia1.nonce1}
+	return &initiatorAct2{nonce1: ia1.nonce1, protocol1: ia1.protocol1}
 }
 
 // AnswerHandshake is used to initiate a responder as a result of receiving
 // message from initiator in the first act of the handshake protocol.
 // The returned responder is in a state ready to execute the second act of the
 // handshake protocol.
-func AnswerHandshake(message *Act1Message) (*responderAct2, error) {
+// The function also validates if both parties run the same protocol.
+func AnswerHandshake(message *Act1Message, protocol string) (*responderAct2, error) {
+	if message.protocol1 != protocol {
+		return nil, fmt.Errorf("unsupported protocol: [%v]", message.protocol1)
```

Comment:
Any chance we can get the sender id into the error message?

Reply:
At this level, sender ID is not confirmed yet and it could be fake. The address is the only thing we are sure about and it is logged.